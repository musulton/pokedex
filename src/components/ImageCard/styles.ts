import Image from "next/image";
import styled from "styled-components";

export const ImageStyled = styled(Image)`
  width: auto;
  height: 130px;
`

interface ImgWrapperProps {
    bgColor: string
}

export const ImgWrapper = styled.div<ImgWrapperProps>`
  margin: 5px;
  width: 150px;
  height: 200px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  overflow: hidden;
  place-self: center;
  border: 2px solid #2C3E50;
  
  padding: 15px;
  background-color: ${p => p.bgColor};
  transition: color .3s, background .3s;
  
  &:hover {
    cursor: pointer;
    color: #fff;
    background-color: #2C3E50;
  }
`

export const TextWrapper = styled.div`
  margin: 15px 0 0 0;
  text-align: center;
`

export const NameText = styled.h3`
  font-size: 16px;
  font-weight: bold;
  margin: 0;
`

export const IdText = styled.h3`
  margin: 0;
  font-size: 14px;
  font-weight: normal;
`
