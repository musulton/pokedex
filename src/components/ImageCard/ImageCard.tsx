import {memo} from "react";
import {useRouter} from "next/router";

import {capitalize, getRandomColor, padWithLeadingZeros} from "@/utils/common";
import {ImageLinkItem} from "@/types";
import {IdText, ImageStyled, ImgWrapper, NameText, TextWrapper} from "@/components/ImageCard/styles";

interface Props {
    imageItem: ImageLinkItem
}

const ImageCard = ({imageItem}: Props): JSX.Element => {
    const router = useRouter()

    return (
            <ImgWrapper
                onClick={() => router.push(`/pokemon/${imageItem?.name}`)}
                bgColor={getRandomColor()}
            >
                <ImageStyled
                    src={imageItem.imageSrc}
                    alt={imageItem.name}
                    width={150}
                    height={150}
                />
                <TextWrapper>
                    <NameText>{capitalize(imageItem.name)}</NameText>
                    <IdText>#{padWithLeadingZeros(imageItem.id, 3)}</IdText>
                </TextWrapper>
            </ImgWrapper>
    )
}

export default memo(ImageCard);
