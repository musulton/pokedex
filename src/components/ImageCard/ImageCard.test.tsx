import {fireEvent, render, screen} from "@testing-library/react";

import ImageCard from "@/components/ImageCard/ImageCard";

const mockPush = jest.fn()
jest
    .mock('next/router', () => ({
        useRouter() {
            return ({
                push: mockPush,
            });
        },
    }))
    .mock("../../utils/common", () => ({
        ...jest.requireActual("../../utils/common"),
        getRandomColor: jest.fn().mockImplementation(() => "white")
    }))

describe("ImageCard", () => {
    const imageItemProps = {
        id: 1,
        imageSrc: "https://avatars.githubusercontent.com/u/45544958?v=4",
        name: "plankton"
    }

    it("should render ImageCard correctly", () => {
        const {container} = render(<ImageCard imageItem={imageItemProps} />)

        expect(container.firstChild).toMatchSnapshot();
    })

    it("should navigate to detail page when image is clicked", () => {
        render(<ImageCard imageItem={imageItemProps} />)
        fireEvent.click(screen.getByText("Plankton"))

        expect(mockPush).toHaveBeenCalledWith(`/pokemon/plankton`)
    })
})
