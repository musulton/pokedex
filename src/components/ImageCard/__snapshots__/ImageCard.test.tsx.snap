// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageCard should render ImageCard correctly 1`] = `
<div
  class="sc-3a13716d-1 bJAiAP"
>
  <img
    alt="plankton"
    class="sc-3a13716d-0 hsXJje"
    data-nimg="1"
    decoding="async"
    height="150"
    loading="lazy"
    src="/_next/image?url=https%3A%2F%2Favatars.githubusercontent.com%2Fu%2F45544958%3Fv%3D4&w=384&q=75"
    srcset="/_next/image?url=https%3A%2F%2Favatars.githubusercontent.com%2Fu%2F45544958%3Fv%3D4&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Favatars.githubusercontent.com%2Fu%2F45544958%3Fv%3D4&w=384&q=75 2x"
    style="color: transparent;"
    width="150"
  />
  <div
    class="sc-3a13716d-2 dyKaBw"
  >
    <h3
      class="sc-3a13716d-3 gyiBgf"
    >
      Plankton
    </h3>
    <h3
      class="sc-3a13716d-4 bvMZfQ"
    >
      #
      001
    </h3>
  </div>
</div>
`;

exports[`should render ImageCard correctly 1`] = `
<div
  class="sc-3a13716d-1 bJAiAP"
>
  <img
    alt="plankton"
    class="sc-3a13716d-0 hsXJje"
    data-nimg="1"
    decoding="async"
    height="150"
    loading="lazy"
    src="/_next/image?url=https%3A%2F%2Favatars.githubusercontent.com%2Fu%2F45544958%3Fv%3D4&w=384&q=75"
    srcset="/_next/image?url=https%3A%2F%2Favatars.githubusercontent.com%2Fu%2F45544958%3Fv%3D4&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Favatars.githubusercontent.com%2Fu%2F45544958%3Fv%3D4&w=384&q=75 2x"
    style="color: transparent;"
    width="150"
  />
  <div
    class="sc-3a13716d-2 dyKaBw"
  >
    <h3
      class="sc-3a13716d-3 gyiBgf"
    >
      Plankton
    </h3>
    <h3
      class="sc-3a13716d-4 bvMZfQ"
    >
      #
      001
    </h3>
  </div>
</div>
`;
